src/compiler/scala/tools/nsc/CompileClient.scala

```diff
@@ -60,7 +60,7 @@ class StandardCompileClient extends HasCompileSocket with CompileOutputCommon {
 object CompileClient extends StandardCompileClient {
   def main(args: Array[String]): Unit = {
     val ok = try process(args) catch { case _: Exception => false }
-    System.exit(if (ok) 0 else 1)
+    if (!ok) throw FatalError("exit(1)")
```

Comment:
`exit(1)` is a somewhat useful error message, no? Especially as methods like this, which are almost certainly going to be used as a main method, will dump the stack. If we're going to give the user a stack trace the one that caused `!ok` may be more useful.

I do dislike that you can't return a status code from a main method.

Reply:
It seems to me that in this case the best way to propagate an actual exception is to just remove the `try`-`catch` block.

I agree that `"exit(1)"` is not a great message, but I'm honestly not sure what else it should be, given that:

1. `process(args)` already prints out any info it wants to show the user
2. I don't actually know what's wrong, other than that it's not "ok"

Reply:
> process(args) already prints out any info it wants to show the user

Right, which is why dying with a stack trace is unfortunate. As I mentioned, best would be to return `Int` and have the JVM handle that if it's used as a main method, but that's not going to happen.

Reply:
An especially evil person could write a `sys.exitIfCurrentlyInTheRealMainMethod` but I think that should be avoided.

Reply:
Another (less trivial) option is to define a trait `Main` which returns an `Int`, such as

```scala
trait Main {
  def doMain(args: Array[String]): Int
}
```

and then have current main methods delegate everything except `sys.exit` to `doMain`. Tools would then call `doMain` instead of `main`, and it would be safe to call `sys.exit` within `main`.

Of course, that still doesn't solve the problem of `scala.tools.cmd.runAndExit`